clients/src/main/java/org/apache/kafka/common/network/KafkaChannel.java

```diff
@@ -350,6 +376,14 @@ public boolean hasBytesBuffered() {
         return transportLayer.hasBytesBuffered();
     }
 
+    /**
+     * @return Time (in milliseconds) at which {@link #prepare()} was last called for this channel; -1 if {@link #prepare()}
+     *         has not yet been called.
+     */
+    public long prepareStartTimeMs() {
+        return prepareStartTimeMs;
```

Comment:
Not sure why we use the start of the prepare time instead of just computing delay from the time the authentication exception was caught.

Reply:
While implementing, I thought it'd be better to send authentication failure responses at fixed time intervals so that we do not leak information about how much time was spent doing the actual authentication. The current implementation delays the response by the configured delay time but also factors in the time we have already spent doing the authentication. We also double the delay if we have already spent the amount of time we want to delay by which makes it more difficult to guess the actual amount of time spent.

Let me know if you think this behavior is reasonable to have.

Reply:
I imagine that typically the delay would be set to quite a small number. As an example, if the delay is 5 ms, the additional delay for different authentication times (i.e. time that the network thread is blocked) would be:
- 4ms => 1ms
- 6ms => 4ms
- 10ms => 0ms
- 21ms =>19ms
The configured delay of 5ms is neither a minimum delay nor a maximum delay. And the delays don't look intuitive either. With a fixed delay starting from the time of the exception, we are not leaking any information beyond what you get with a configured delay of zero? If we really want to avoid leaking information, perhaps we need to randomize.

To me, `prepareStartTimeMs` doesn't represent a time that we can easily document when documenting the config. It is not the time taken for authentication, it is the time at which the current packet processing started. Perhaps if you can update `KafkaConfig.FailedAuthenticationDelayMsDoc`, it will be easier to see if this is better than a fixed delay.

Reply:
Another way to think about the configuration is that broker will send out responses for failed authentication after a fixed amount of time has elapsed since the processing of the packet began. For example, if the delay is set to 5ms, the broker will send out the failed authentication response approximately 5ms after the packet processing started. If more than 5ms have already elapsed since we started processing the packet, the response will be sent after 10ms have elapsed (and similarly after 20ms, 40ms, 80ms and so on).

I agree that with a fixed delay, we do not leak more information than without the delay. I thought the delay itself could be one way to implement a safeguard so that we do not leak this timing information. If you think we need some more thought on this, I can revert to using the fixed delay for now and we can think about how to implement safeguard against leaking timing information in a subsequent PR / KIP. Let me know what you think.

Reply:
Some thoughts:

1. If it's tricky to do the more sophisticated version, we can stick to fixed delays to start with. It's then up to authentication implementations to ensure that they are not prone to timing attacks. @rajinisivaram do you know if SASL/SCRAM is resistant to timing attacks?

2. From what @dhruvilshah3 told me, the approach used here is similar to what SSH does, so there's precedent.

3. @rajinisivaram, I was a bit confused about your statement that values would typically be very low and using 5ms as an example. The KIP states that the default would be 100ms, right?